### PR TITLE
Add `--app-name` option to `mlflow server` to allow specifying mlflow app name

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #7757

## What changes are proposed in this pull request?

Adds an `--app-name` option to `mlflow server` to allow specifying the mlflow app name.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

```
> mlflow server --app-name custom_app
[2023-03-07 15:23:46 +0900] [26442] [INFO] Starting gunicorn 20.1.0
[2023-03-07 15:23:46 +0900] [26442] [INFO] Listening at: http://127.0.0.1:5000 (26442)
[2023-03-07 15:23:46 +0900] [26442] [INFO] Using worker: sync
[2023-03-07 15:23:46 +0900] [26453] [INFO] Booting worker with pid: 26453
[2023-03-07 15:23:46 +0900] [26454] [INFO] Booting worker with pid: 26454
[2023-03-07 15:23:46 +0900] [26455] [INFO] Booting worker with pid: 26455
[2023-03-07 15:23:46 +0900] [26456] [INFO] Booting worker with pid: 26456
Using mlflow_test_plugin.app
Using mlflow_test_plugin.app
Using mlflow_test_plugin.app
Using mlflow_test_plugin.app
```

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/__init__.py [hunk 1]
```diff
@@ -107,22 +107,16 @@ def serve():
...
+def _find_app(app_name: str) -> str:
+    """Find the entrypoint for the given app name."""
+    apps = entrypoints.get_group_all("mlflow.app")
+    for app in apps:
+        if app.name == app_name:
+            return f"{app.module_name}:{app.object_name}"
+
+    raise MlflowException(
+        f"Failed to find app: {app_name}. Available apps: {[a.name for a in apps]}"
+    )
```
Comment (WeichenXu123) [inline mlflow/server/__init__.py:119]: Why removing this part ?
```
    # Cannot install more than one
    if len(apps) > 1:
        raise MlflowException(
            "Multiple server plugins detected. "
```

Is it impossible to occur ?

File: docs/source/plugins.rst [hunk 2]
```diff
@@ -218,10 +218,10 @@ plugin:
      - `DummyEvaluator <https://github.com/mlflow/mlflow/blob/branch-1.23/tests/resources/mlflow-test-plugin/mlflow_test_plugin/dummy_evaluator.py>`_.
    * - [Experimental] Plugins for custom mlflow server flask app configuration `mlflow.server.app <https://github.com/mlflow/mlflow/blob/v2.2.0/mlflow/server/__init__.py#L30>`_.
      - mlflow.app
-     - The entry point name (e.g. ``mlflow_test_plugin:app``) specifies a customized flask application. This can be useful for implementing
+     - The entry point name (e.g. ``custom_app=mlflow_test_plugin.app:app``) specifies a customized flask application. This can be useful for implementing
```
Comment (WeichenXu123) [inline docs/source/plugins.rst:221]: We can describe it more clear: in the example ``custom_app=mlflow_test_plugin.app:app``, the `custom_app` is the --app-name value.

**Review by WeichenXu123 — APPROVED** [review]

File: docs/source/plugins.rst [hunk 3]
```diff
@@ -216,12 +216,12 @@ plugin:
...
-     - The entry point name (e.g. ``mlflow_test_plugin:app``) specifies a customized flask application. This can be useful for implementing
-       request hooks, custom logging and custom flask configurations. The plugin must import `mlflow.server.app` (e.g. ``from mlflow.server import app``) and may add custom configuration, middleware etc. to the app.
+     - The entry point ``<app_name>=<object_reference>`` (e.g. ``custom_app=mlflow_test_plugin.app:app``) specifies a customized flask application. This can be useful for implementing
+       request hooks for authentication/authorization, custom logging and custom flask configurations. The plugin must import `mlflow.server.app` (e.g. ``from mlflow.server import app``) and may add custom configuration, middleware etc. to the app.
        The plugin should avoid altering the existing application routes, handlers and environment variables to avoid unexpected behavior.
-       Users who install the example plugin will have a customized flask application.
+       Users who install the example plugin will have a customized flask application. To run the customized flask application, use ``mlflow server --app-name <app_name>``.
```
Comment (harupy) [inline docs/source/plugins.rst:224]: @WeichenXu123 Improved the doc.

File: docs/source/plugins.rst (hunk 2 above)
Comment (harupy) [reply docs/source/plugins.rst]: Updated

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**